vermouth/tests/test_name_moltype.py

```diff
@@ -0,0 +1,151 @@
+        :func:`hypothesis.strategies.lists`.
+    molecule_kwargs: dict
+        Arguments to pass to :func:`random_molecule`.
+
+    Returns
+    -------
+    hypothesis.searchstrategy.lazy.LazyStrategy
+        A list of molecules and a list of moltype names corresponding to these
+        molecules as they are expected to be assigned by
+        :class:`vermouth.processors.name_moltype.NameMolType`.
+    """
+    if max_size is not None and max_size < min_size:
+        raise ValueError('max_size ({}) must be greater than min_size ({}), or None.'
+                         .fornat(max_size, min_size))
+
+    # We first generate the list of expected moltypes, we can then use it as
+    # keys to draw the molecules with shared strategies. The moltypes must
+    # match what NameMolType would generate, so they must have the form
+    # molecule_0, molecule_1, molecule_x, where the index is increasing based
+    # on the first occurence of the molecule type.
+    # pre_moltypes will be the moltype names before we rename them based on
+    # their order.
+    n_moltypes = draw(st.integers(min_value=1, max_value=max_moltypes))
+    unique_pre_moltypes = range(n_moltypes)
+    pre_moltypes_sampled = draw(st.lists(st.sampled_from(unique_pre_moltypes),
+                                         min_size=min_size, max_size=max_size))
+    if not pre_moltypes_sampled:
+        return [], []
+
+    moltype_index = 0
+    pre_to_moltype = {pre_moltypes_sampled[0]: 'molecule_{}'.format(moltype_index)}
+    moltypes = []
+    for pre_moltype in pre_moltypes_sampled:
+        if pre_moltype not in pre_to_moltype:
+            moltype_index += 1
+            pre_to_moltype[pre_moltype] = 'molecule_{}'.format(moltype_index)
+        moltypes.append(pre_to_moltype[pre_moltype])
+
+    sharing = defaultdict(lambda : draw(random_molecule(**molecule_kwargs)))
+    molecules = [sharing[moltype] for moltype in moltypes]
+
+    # So far we made sure that molecules with a different moltype are drawn
+    # separately. Though, two different draws may result in equal molecules,
+    # especially as molecules are reduced whn shrinking the draws.
+    representatives = [
+        next(group)[0]
+        for _, group in itertools.groupby(
+                zip(molecules, moltypes), key=lambda x: x[1])
+    ]
+    assume(all(not mol1.share_moltype_with(mol2)
+               for mol1, mol2 in itertools.combinations(representatives, 2)))
+
+    # Each molecule must be its own instance as molecules are modified in place.
+    molecules = [molecule.copy() for molecule in molecules]
+
+    return [molecules, moltypes]
+
+
+@pytest.mark.parametrize('deduplicate', (True, False))
+@given(mols_and_moltypes=molecules_and_moltypes(
+    max_size=4, max_moltypes=3,  molecule_kwargs={'max_nodes': 3, 'max_meta': 3},
+))
+def test_name_moltype(mols_and_moltypes, deduplicate):
+    molecules, moltypes = mols_and_moltypes
+
+    system = System()
+    system.molecules = molecules
```

Comment:
TODO: make sure system.molecules is a property that deals with e.g. forcefields

Reply:
It would be a very good idea indeed. Though, a property can only detect cases where the attribute is reassigned, not when it is mutated.

Reply:
True, but it would help. One option would be to make `system.molecules` a property that refers to a `system._molecule` that is not a list, but a list subclass.

Reply:
Good idea. Write an issue.